src/agents/supervisor/agent.py

```diff
@@ -24,48 +26,35 @@ class SupervisorAgent:
 
     def __init__(self, model: IModel, members: list[str]):
         self.model = model
+        self.options = [FINALIZER] + members
+        self.parser = self._create_parser()
+        self.supervisor_chain = self._create_supervisor_chain()
 
-        options: list[str] = [FINALIZER] + members
-        function_def = {
-            "name": "assign_and_route",
-            "description": "Assign subtasks to agents.",
-            "parameters": {
-                "type": "object",
-                "properties": {
-                    "subtasks": {
-                        "type": "array",
-                        "items": {
-                            "type": "object",
-                            "properties": {
-                                "description": {"type": "string"},
-                                "assigned_to": {"enum": options},
-                            },
-                            "required": ["description", "assigned_to"],
-                        },
-                    },
-                    "next": {"type": "string", "enum": options},
-                },
-                "required": ["subtasks", "next"],
-            },
-        }
+    def _create_parser(self) -> PydanticOutputParser:
+        class RouteResponse(BaseModel):
+            next: Literal[*self.options] | None = Field(  # type: ignore
+                description="next agent to be called"
+            )
 
+        return PydanticOutputParser(pydantic_object=RouteResponse)
+
+    def _create_supervisor_chain(self) -> RunnableSequence:
         supervisor_system_prompt = ChatPromptTemplate.from_messages(
             [
                 ("system", SUPERVISOR_ROLE_PROMPT),
                 MessagesPlaceholder(variable_name="messages"),
                 ("assistant", "Subtasks: {subtasks}"),
                 ("system", SUPERVISOR_TASK_PROMPT),
             ]
-        ).partial(options=str(options), members=", ".join(options), finalizer=FINALIZER)
-
-        self.supervisor_chain = (
-            supervisor_system_prompt
-            | model.llm.bind_functions(
-                functions=[function_def], function_call="assign_and_route"
-            )
-            | JsonOutputFunctionsParser()
+        ).partial(
+            options=str(self.options),
+            members=", ".join(self.options),
+            finalizer=FINALIZER,
+            output_format=self.parser.get_format_instructions(),
         )
 
+        return supervisor_system_prompt | self.model.llm  # type: ignore
```

Comment:
Why can this function either return a ChatPromptTemplate or model.llm?

Reply:
The sign `|` is pipe sign of the langchain where it used to create a langchain chain. It is not a OR sign :)